[AMD] added slicing `ttg.async_copy_global_to_local`

ravil-mobile opened the pull request:
```bash
> TRITON_HIP_USE_BLOCK_PINGPONG=1 TRITON_HIP_USE_ASYNC_COPY=1 pytest -s -v op_tests/triton_tests/test_gemm_afp4wfp4.py

...

================================================================================================= warnings summary =================================================================================================
op_tests/triton_tests/test_gemm_afp4wfp4.py::test_gemm_afp4_wfp4[dtype0-1024-1024-1024]
  /home/rdorozhi/work/aiter/op_tests/triton_tests/test_gemm_afp4wfp4.py:92: UserWarning: Attempting to use hipBLASLt on an unsupported architecture! Overriding blas backend to hipblas (Triggered internally at /var/lib/jenkins/pytorch/aten/src/ATen/Context.cpp:328.)
    return torch.mm(x_f32, w_f32).to(dtype)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================================================== 74 passed, 1 warning in 65.59s (0:01:05) =====================================================================================


```

AlexAUT committed:
[FA] 4-stage FA pipeliner

4-stage FA experiment

Cluster assignment

AlexAUT committed:
[FA] Add FA scripts

AlexAUT committed:
[FA] Place cvt layout in the same stage and cluster as LocalLoad so canonicalize can fold it

AlexAUT committed:
[ASYNC_COPY] Add env var to bypass permute, only works if the load dim is contiguous

AlexAUT committed:
[FA] Do not combine AsyncWaits to have a barrier in front of each memory ops cluster

AlexAUT committed:
[ASYNC_COPY] Remove MemoryEffect of BufferLoadToLocal to avoid implicit barrier from Membar

AlexAUT committed:
[FA] Compute max before mul QK_SCALE to fold sub into fma

AlexAUT committed:
[FA] Added 2 extra clusters to have async_waits in front of memory clusters

AlexAUT committed:
[FA] Place LocalLoads before AsyncCopies

AlexAUT committed:
[FA][ASYNC_COPY] Force vec=8 for shared encodings to avoid 32bit direct to lds loads

AlexAUT committed:
[FA] Place dots at the top of clusters

AlexAUT committed:
[FA] Split 4-stage clusters into 8 clusters to better controll the order in the loop

AlexAUT committed:
[FA] Revert order change in SM clusters

zhanglx13 committed:
[FA] Set vecSize=nonKDim for V shared layout to avoid bank conflicts

I'll submit a PR upstream later.

AlexAUT committed:
[FA] Removed old vectorSize workaround

AlexAUT committed:
[FA] Revert "Place AsyncWait at the top of schedule"

This reverts commit 718ee20be3bcdeb75057d210f659697f34995653.

jungpark-mlir committed:
[FA][PINGPONG] Add support for FAv3 pingpong.

Initial support over already arranged ops.

AlexAUT committed:
[FA][PINGPONG] Allow block pingpong with num_stages==4

jungpark-mlir committed:
[GEMM] Add combine dot_scaled and addF

AlexAUT committed:
[FA][PINGPONG] Bail out if async wait count != 2

AlexAUT committed:
[FA] Do not pipeline second loop (causal)

AlexAUT committed:
[GEMM] Do not swizzle the scale

AlexAUT committed:
[FA] Split FourStagePipeliner to separate file and do very basic selection based on the loop. This is not meant as a permanent solution just to make this branch useable for other workloads

vgokhale committed:
Add layout conversion pass optim at the end

jungpark-mlir committed:
Initial commit to enable pingpong for dot_scaled with mxfp4

Computation part interleaves mfma and ds_read
Placed extra conditional barrier to overlap computation part
and buffer_load part. Dot slicing by plognjen at https://github.com/plognjen/triton/tree/slice_dot_scaled
requires vmcnt fix to achieve full performance.

jungpark-mlir committed:
Fix to the gemm pingpong.

Fix incorrect condition to choose enable transforms.
Fix missing tokens to the local_load

jungpark-mlir committed:
Add restriction to dot_scaled pingpong.

Only enable for 256x256x256 tilesize

AlexAUT committed:
Revert "[AMD] Use v_permlane to optimize MFAM to linear layout on GFX950 (#6744)"

This reverts commit 4ecc496ed82c0e2cfd721c6ee71d0b468975c7f9.

AlexAUT committed:
Revert "[BACKEND] bump to llvm/llvm-project@3c709802d31b (#6754)"

This reverts commit f3076b136c62ba95423a1d49f228376bc3da800f.

AlexAUT committed:
Revert because no longer needed: "[ASYNC_COPY] Remove MemoryEffect of BufferLoadToLocal to avoid implicit barrier from Membar"

This reverts commit 012793ae2809884143a4e8dfdbdfd7f72c3382d0.

raikonenfnu committed:
[AMD] Enable async pingpong for F16 GEMMs (#796)

* [AMD] Generalize PingPong to have different type of Load/Store Ops

This main motivation behind this commit is to add support for PingPong
with AsyncOps. In order to accomplish that we made these changes:
- Fork "determineDotMemoryOps" to "determineDotAsyncMemoryOps" that handles async memory ops.
- Refactor validation and pruning of memory ops to "pruneDotMemoryOps"
  S.T we can have clean interface for it's async memory ops counterpart
  "pruneAsyncDotMemoryOps".
- Plumb "useBlockPingpong" into StreamPipeliner S.T it can adjust AsyncWait
  stage/cluster to hoist first AsyncWait and allow set AsyncWait towards
  the end of the loop to make it easier for 4 PP cluster to move it
  before the 3rd dot-slice / 2 s_barrier before localLoads
  this is to ensure no race conditions.
- Add check to enable handling of dotSOps (dot scaled) VS dotOps (dot)

Signed-off-by: Stanley Winata <stanley.winata@amd.com>
Co-authored-by: Alexander Weinrauch <alexander.weinrauch@amd.com>

raikonenfnu commented:
can we have an option to **not** slice global op even if we slice dots?

ravil-mobile committed:
[AMD] add slicing `async-copy-local-to-global`

ravil-mobile commented:
> can we have an option to **not** slice global op even if we slice dots?
if you set `TRITON_HIP_USE_ASYNC_COPY=0` then only `tt.Dot`s or `tt.scaleDot` are going to be sliced

raikonenfnu commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -709,25 +733,72 @@ LogicalResult Pingponger::genLocalSliceHelper(OpBuilder &builder, Value v,
+          triton::amdgpu::OpIdxAttr::getMnemonic(),
+          triton::amdgpu::OpIdxAttr::get(ctx,
+                                         dotOperandResEncoding.getOpIdx()));
+      dotOperandResEncoding.getOpIdx();
```
dead code?

raikonenfnu commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -709,25 +733,72 @@ LogicalResult Pingponger::genLocalSliceHelper(OpBuilder &builder, Value v,
+
+    builder.setInsertionPointAfter(origMemDesc);
+    auto sliceIdAttr = mlir::IntegerAttr::get(intType, i);
+    if (refineOrigSubview) {
```
can we guard this to only work from the FP4/dotScaled if not already? I think right now this is guarded by:
```
  auto dotOperandResEncoding =
      dyn_cast<ttg::DotOperandEncodingAttr>(resEncoding);
  const bool refineOrigSubview = dotOperandResEncoding != nullptr;
```

jungpark-mlir replied:
Yeah, I'd add `needCopySliced` arg to genLocalSliceHelper(), so we can choose from the different transforms.

raikonenfnu replied:
but the fp16 pattern we are looking at will now be affected right, which may not be intended

jungpark-mlir replied:
Yeah, I mean. f16 pattern to call this function with needCopySliced=false then it's not affected? maybe I'm missing something.

raikonenfnu replied:
ah misread your commen there, but yeah if we have that flag that'd work!

ravil-mobile replied:
Added `needCopySliced` arg to `genLocalSliceScales`

raikonenfnu commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -772,6 +843,389 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
+    // infer the sliced shape
+    triton::gpu::MemDescSubviewOp subViewSlice = subViews[0];
+    auto origShape = subView.getType().getShape();
+    auto slicedShape = subViewSlice.getType().getShape();
```
Just double checking, is the subViewSlice coming from
```
      Value newOrigSmem = builder.create<ttg::MemDescSubviewOp>(
          origMemDesc.getLoc(), subviewDescType, origMemDesc, offsetsVal);
```

ravil-mobile replied:
We slice 2 subviews for each tile - i.e., 1) one which proceeds `local_load` immediately (usually comes from an arg for `scf::ForOp` and 2) the subview from a previous loop iteration (used by either `local_store` or `async_copy_local_to_global`. The former is called `newSmem`; the later is `origMemDesc`

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -82,10 +106,13 @@ class Pingponger {
   void transformOnePPClusters(OpBuilder &builder, Location loc);
   LogicalResult transformFourPPClusters(OpBuilder &builder, Location loc);
   LogicalResult transformTwoPPClusters(OpBuilder &builder, Location loc);
-  LogicalResult transformFAv3(OpBuilder &builder, Location loc);
```
Please don't delete this

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -989,47 +1449,6 @@ LogicalResult Pingponger::transformTwoPPClusters(OpBuilder &builder,
   return success();
 }
 
-// Fixme : document the scheduling.
```
Don't delete this function

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -1103,10 +1521,10 @@ void Pingponger::addAsymmetricSyncToLoop(OpBuilder &builder, Location loc) {
+  if (numStages != 2) {
     std::stringstream message;
-    message << "All ping pong scheduling requires 2 or 4 stages. Found "
-            << numStages << " stages";
```
FAv3 needs numstage==4, still need this.

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -1136,23 +1555,22 @@ void Pingponger::getDotPingponged() {
   });
 
-  // Fixme : use proper condition to identify FAv3
-  if (numStages == 4 && dotOps.size() == 2) {
```
don't delete this

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -772,6 +843,389 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
+  origAsyncWait->replaceAllUsesWith(newAsyncWait);
+  origAsyncWait->erase();
+
+  // collect all refined lodalLoads
```
typo

jungpark-mlir commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -772,6 +843,389 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
+      auto extractedMask = extract(slicedMaskType, newMask, offsetAttr);
+      auto extractedOther = extract(slicedOtherType, newOther, offsetAttr);
+
+      auto newAsyncCopy = builder.create<ttg::AsyncCopyGlobalToLocalOp>(
```
Can we keep pointers to the new copy slices as well?
We may want to relocate everything.

ravil-mobile replied:
Added attributes to `newAsyncCopy` - i.e., `DotOpIdx` and `sliceId`

jungpark-mlir reviewed (changes requested):
FAv3 related code is accidentally deleted by the merge, and some few changes.

ravil-mobile commented:
@jungpark-mlir, @raikonenfnu. Thanks for your comments! I addressed all of them.

joviliast committed:
[AMD] Add an option to force async copy overlapping

Signed-off-by: Ilya Veselov <iveselov.nn@gmail.com>

AlexAUT committed:
[ASYNCCOPY] Simplify swizzling calculations to get better codegen from
the backend

antiagainst committed:
[AMD] Use composition to swap columns for mfma like store layout (#6844)

This commit improves how we create the mfma-like layout for
optimizing global store by using linear layout composition.
Along the way fixes a few implemenation issues.

---------

Co-authored-by: Yi Qian <yi.qian@amd.com>

jungpark-mlir committed:
Add initial support for skinny mxfp gemm

Overlapping buffer_load and local_load+dot

jungpark-mlir committed:
add AB load separated pingpong for skinny gemm.

raikonenfnu commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -772,6 +846,390 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
+    auto sizePerThread = encoding.getSizePerThread();
+    SmallVector<unsigned> threadPerWarp(warpsPerCTA.size(), 0);
+    for (size_t dim = 0; dim < numDims; ++dim) {
+      threadPerWarp[dim] =
```
I have a silly Q for educational purposes, looks like we are only updating the threadsPerWarp here, does it not make more sense to updated the sizePerThread in the newEncoding?

ravil-mobile replied:
No, we want to preserve the number of elements per thread (i.e., each holds `16` consecutive elements of a tensor). We just want to change the layout of threads; thus, to change which part of a tensor is held by CTA.

Each CTA holds `64x128` tile of a tensor in the following example 

```mlir
#blocked = #ttg.blocked<{sizePerThread = [1, 16], threadsPerWarp = [8, 8], warpsPerCTA = [8, 1], order = [1, 0]}>
```
But if we change `threadsPerWarp` to `[32,2]` then a CTA holds `256x32` tile of a tensor. 

```mlir
#blocked3 = #ttg.blocked<{sizePerThread = [1, 16], threadsPerWarp = [32, 2], warpsPerCTA = [8, 1], order = [1, 0]}>
```

In both cases, a thread holds 16 consecutive elements which determines the width of load instructions

raikonenfnu replied:
Thanks for the answer, that's really interesting! A follow up question, if we change the which part of tensor the CTA will hold, wouldn't we need to have an extra global read (or potentially fused but wider global read) to get that part of tensor?

ravil-mobile replied:
ConverLayout results in some machine ops if it cannot be fully optimized (propagated till the top of a function). If I remember correctly, the layout change is going to happen in LDS. In out case, it is going to be optimized

raikonenfnu replied:
Nice! didn't realize this was pre-layout propagation phase, but this makes a lot of sense, thanks :)

raikonenfnu commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -772,6 +846,390 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
+        auto elemType = tensor.getType().getElementType();
+        newType = RankedTensorType::get(origShape, elemType, newEncoding);
+        newTensor =
+            builder
+                .create<ttg::ConvertLayoutOp>(tensor.getLoc(), newType, tensor)
+                .getResult();
+        slicedTensorType =
+            RankedTensorType::get(slicedShape, elemType, newEncoding);
```
Another silly Q, I know amdgpu.extract_slice src and dst layout need to match, is this the main purpose of ttg.convert_layout to set up layout for stuff like extract_slice where the layout is for sure to change? On top of that, does that mean ttg.convert_layout for slices will most likely not have a matching layout to it's shape?

ravil-mobile replied:
dst layout is determined by the source layout. The problem comes from the following. Let's assume we want to slice `256x128` tensor into 4 pieces of `256x32` tiles. Let's also assume that the original layout is 

```mlir
# orig-layout
#blocked = #ttg.blocked<{sizePerThread = [1, 16], threadsPerWarp = [8, 8], warpsPerCTA = [8, 1], order = [1, 0]}>
```

which holds `64x128` tiles per CTA. `ExtractSliceOp` is the CTA level op - i.e., it can slice a tensor if a new size  is proportional to the CTA tile - i.e.,`64x128` in our case. Therefore, we cannot apply `ExtractSliceOp` to our tensor with `orig-layout`. Thus, we change the source layout to a new one - i.e., 


```mlir
# new-layout
#blocked3 = #ttg.blocked<{sizePerThread = [1, 16], threadsPerWarp = [32, 2], warpsPerCTA = [8, 1], order = [1, 0]}>
```
which has its CTA-level slice equaled to `256x32`. Now we can slice `256x128` tensor into 4 pieces. (note: it was determined by dot-slicing)

raikonenfnu commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -772,6 +846,390 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
+        ctx, sizePerThread, threadPerWarp, warpsPerCTA, encoding.getOrder(),
+        encoding.getCTALayout());
+
+    auto converTensor = [&](mlir::TypedValue<mlir::RankedTensorType> tensor) {
```
NIT: I think it's missing a `t` for `convert` i.e `convertTensor`. Any thoughts on the name `convertLayout`?

ravil-mobile replied:
Thanks!

raikonenfnu commented on third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp:
```diff
@@ -772,6 +846,390 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
+        encoding.getCTALayout());
+
+    auto converTensor = [&](mlir::TypedValue<mlir::RankedTensorType> tensor) {
+      RankedTensorType newType = nullptr;
+      Value newTensor = nullptr;
+      RankedTensorType slicedTensorType = nullptr;
```
NIT: IIRC, RankedTensorType/Type/Value will default to null even if you don't explicitly set nullptr

ravil-mobile replied:
You are right. I just like to be explicit in the code about initialization of all local variables (old habit).

ravil-mobile committed:
[AMD] Improved CanonicalizePointers for ExtractSlice

antiagainst committed:
Revert "Revert "[AMD] Use v_permlane to optimize MFAM to linear layout on GFX950 (#6744)""

This reverts commit f6065b93e96e1cdb5137e27f7b8f21199399cf54.

jungpark-mlir committed:
Code cleanup

avoid wrongly enabled.

jungpark-mlir committed:
Add skinny pingpong transform

Requirement to enable the transform
: mxfp4, 128x128x512 tile size, async_copy, num_stages=2, num_warps=8

AlexAUT committed:
[FA] Disable pipelining for causal loop

plognjen committed:
[AMD] Add a Concat op to AMDGPU dialect (#6590)

The "concat" operation combines a list of source n-dimensional tensors
into a single larger destination tensor.

All source tensors must have the same shape, element type, and encoding.
The concatenation dimension is inferred from the source and destination
shapes provided by the user.
For example, two tensors of shape 64x128 can produce a destination shape
of 128x128,
indicating concatenation along dimension 0; or 64x256, indicating
concatenation along dimension 1.

Generally, source tensors passed as op arguments can be arranged into
the resulting shape in multiple ways.
For example, given four tensors of shape 64x64:
  concat s0<64x64>, s1<64x64>, s2<64x64>, s3<64x64> -> <128x128>

They can be laid out in different configurations within the result
tensor:
   1) s0 s1 
       s2 s3  

   2) s0 s2
        s1 s3

From a logical tensor perspective, the source tensors are treated as
elements of a tensor of tensors.
In other words, the 1-D array of input tensors is conceptually reshaped
into an n-D grid.
The semantics of this op assume a row-major order (or its n-D
generalization),
meaning the fastest-varying dimension is filled first, and the
slowest-varying dimension is filled last.
In the example above, this corresponds to layout 1).

The source and destination tensors must have identical linear layouts at
the CTA tile level.
That is, all base vectors for input dimensions must match, except for
the register input dimension.
The register basis must align on the subset that defines the logical
tensor shape of a single CTA tile.

This ensures that the concatenation is a no-op, meaning no data
rearrangement among threads is required
to assemble the destination tensor with the given shape and layout.
However, the order of CTA tiles within the layout does not need to match
between source and destination layouts.
It is the responsibility of the op's lowering logic to handle this
correctly.

This op is designed to work on logical tensors directly, avoiding the
need for complex layout reinterpretation or reshaping.
For example, the `tt.join` operation only supports concatenation along
the innermost dimension,
and requires that the resulting innermost dimension provide 2 elements
per thread, distributed across registers.
In contrast, this `concat` op imposes no constraints on the
concatenation dimension or the size of dimensions.

---------

Co-authored-by: Ognjen Plavsic <plognjen@amd.com>
Co-authored-by: Lei Zhang <antiagainst@gmail.com>

AlexAUT committed:
WA for incorrect strides in subview

joviliast committed:
[AMD] Add an option to force async copy overlapping

Use `TRITON_HIP_ASYNC_COPY_OVERLAP=1` env to enable async copy overlap

ravil-mobile committed:
Merge branch 'shared/triton-gfx950-launch' into shared/triton-gfx950-launch-update-rebase

ravil-mobile committed:
[AMD] improved subviewing for async-copy-local-to-global
